File: systems-for-design-and-implementation/labs/exam/TicTacToeNetworking/src/tictactoe/network/utils/AbstractConcurrentServer.java
package tictactoe.network.utils;

import java.net.Socket;

/**
 * Wrapper around Threaded Socket
 */
public abstract class AbstractConcurrentServer extends AbstractServer
{
    public AbstractConcurrentServer(int port)
    {
        super(port);
        System.out.println("Concurrent AbstractServer");
    }

    protected void processRequest(Socket client)
    {
        Thread tw = createWorker(client);
        tw.start();
    }

    protected abstract Thread createWorker(Socket client);
}

File: systems-for-design-and-implementation/labs/lab3/LibraryNetworking/src/library/network/utils/AbstractConcurrentServer.scala
package library.network.utils

import java.net.Socket

abstract class AbstractConcurrentServer(port: Int) extends AbstractServer(port)
{
    println("Concurrent AbstractServer")

    protected def processRequest(client: Socket)
    {
        val tw: Thread = createWorker(client)
        tw.start()
    }

    protected def createWorker(client: Socket): Thread
}